lib/deepcrawl.js

```diff
@@ -0,0 +1,137 @@
+  _ = require('lodash');
+
+class DeepCrawl {
+  constructor(cfg) {
+    cfg = cfg || {};
+
+    this.apiId = cfg.apiId;
+    this.apiKey = cfg.apiKey;
+    this.getSessionToken = cfg.getSessionToken;
+    this.accountId = cfg.accountId;
+
+    this.schema = cfg.schema || schema;
+    this.version = this.schema.version;
+    this.baseUrl = this.schema.baseUrl;
+  }
+
+  createEndpoints (schema) {
+    const API = {};
+    for (const resource in schema.resources) {
+      API[resource] = {};
+    }
+    return API;
+  }
+
+  generateRequest (url, options, method) {
+    // first, get the sessionToken and generate the request options and headers
+    return BPromise.resolve(options.sessionToken || this.getSessionToken())
+      .then((sessionToken) => {
+        const requestOpts = {
+          json: true,
+          headers: {
+            'X-Auth-Token': sessionToken
+          }
+        };
+
+        if (method === 'get') {
+          return needle.getAsync(url, requestOpts)
+            .then(this.handleResponse);
+        }
+
+        // convert all camelCased options to _ for deepcrawl
+        const data = humps.decamelizeKeys(options);
+        return needle[`${method}Async`](url, data, requestOpts)
+          .then(this.handleResponse);
+      });
+  }
+
+  generateMethod (requiredFields, url, method) {
+    return (options) => {
+      options = options || {};
+      if (!options.sessionToken && !this.getSessionToken) {
+        throw new Error(`"sessionToken" or "getSessionToken" method required`);
+      }
+
+      requiredFields.forEach((field) => {
+        if (!options[field] && !this[field]) {
+          throw new Error(`${field} is required`);
+        }
+        // if the url contains {field}, replace {field} with the passed
+        // in value
+        if (url.indexOf(`{${field}}`) > -1) {
+          url = url.replace(`{${field}}`, (options[field] || this[field]));
+        }
+      });
+
+      return this.generateRequest(url, options, method);     
+    };
+  }
+
+  addMethodsToEndpoints (API, schema) {
+    for (const resource in API) {
+      const resourceUrl = this.baseUrl + schema.resources[resource].route,
+        requiredFields = _.clone(schema.resources[resource].requiredFields || []),
+        id = schema.resources[resource].id;
+
+      for (const action in schema.resources[resource].actions) {
+        let actionRequiredFields = requiredFields,
+          actionUrl = resourceUrl;
+        // check to see whether the resource action has extra data associated with it
+        const isString = typeof schema.resources[resource].actions[action] === 'string';
+        // the following actions should all have an "id" field associated with them
+        // we will append the id field to the required fields for the action, and
+        // add /{id} to the url, which will be replaced later with the passed in
+        // field
+        if (action === 'read' || action === 'update' || action === 'delete') {
```

Comment:
Well I guess you still want to append the id even if requiredFields is empty, but the regex is still a nicer check.